### PR TITLE
GPTNeoForQuestionAnswering

### PR DESCRIPTION
**Pull request by peter-sk**
# What does this PR do?

Adds QA support for GPT Neo.

Includes PR #23030.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

@ArthurZucker
@younesbelkada

### CODE REVIEW

**Comment by peter-sk** [comment]
One more to go (for now).

**Comment by peter-sk** [comment]
@sgugger @younesbelkada  - as Arthur is on holdiays 👍

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this, I guess you forgot to remove the GPT2 part as it is already done here: https://github.com/huggingface/transformers/pull/23030 - could you remove the changes relative to GPT2?
Otherwise the GPT-Neo code looks very nice to me!

**Comment by peter-sk** [comment]
@younesbelkada 
I merged with main to isolate the GPT Neo specific parts. Now everything seems to work fine.

**Review by younesbelkada — COMMENTED** [review]
Thanks for iterating, for some reason some CI tests are failing, can you try to rebase/main with the main branch? Then we should be good for merging this

**Comment by peter-sk** [comment]
rebased - let's see whether it helps

**Comment by peter-sk** [comment]
@younesbelkada merging helped - ready to merge 👍

**Review by younesbelkada — APPROVED** [review]
Awesome! Thanks a lot for your contribution

**Review by sgugger — COMMENTED** [review]
Thanks a lot, just one small comment and this should be good to merge.

**Review by amyeroberts — APPROVED** [review]
Thanks for adding! ❤️

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 1]
```diff
@@ -1012,3 +1014,104 @@ def forward(
...
+        sequence_output = outputs[0]
+
+        logits = self.qa_outputs(sequence_output)
+        start_logits, end_logits = logits.split(1, dim=-1)
+        start_logits = start_logits.squeeze(-1).contiguous()
+        end_logits = end_logits.squeeze(-1).contiguous()
+
+        total_loss = None
+        if start_positions is not None and end_positions is not None:
+            # If we are on multi-GPU, split add a dimension
```
[the remark this answers is not in the document]
Comment (peter-sk) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:1091]: Any guidance on how to do that? I am not very accustomed to accelerate (yet). Thanks in advance!

[the remark this answers is not in the document]
Comment (peter-sk) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:1091]: Also, the same applies for GPT2ForQuestionAnswering, GTPJForQuestionAnswering etc.

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:1091]: Just move `start_position` and `end_positions` to the device of `start_logits` and `end_logits`.

[the remark this answers is not in the document]
Comment (peter-sk) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py:1091]: Like this, @sgugger?

**Review by sgugger — APPROVED** [review]

**Pull request merged**